File: src/lib/cooking-helpers.ts
import { Ingredient, Recipe } from '@/types';

// Scale recipe ingredients based on new serving size
export function scaleRecipe(recipe: Recipe, newServings: number): Ingredient[] {
  const ratio = newServings / recipe.servings;

  return recipe.ingredients.map((ing) => {
    const originalAmount = parseFloat(ing.amount) || 0;
    const scaledAmount = originalAmount * ratio;

    // Format nicely (avoid ugly decimals)
    let formattedAmount: string;
    if (scaledAmount === Math.floor(scaledAmount)) {
      formattedAmount = scaledAmount.toString();
    } else if (scaledAmount < 1) {
      // Convert to fractions for small amounts
      formattedAmount = toFraction(scaledAmount);
    } else {
      formattedAmount = scaledAmount.toFixed(1).replace(/\.0$/, '');
    }

    return {
      ...ing,
      amount: formattedAmount,
    };
  });
}

// Convert decimal to fraction string
function toFraction(decimal: number): string {
  const fractions: [number, string][] = [
    [0.125, '1/8'],
    [0.25, '1/4'],
    [0.333, '1/3'],
    [0.375, '3/8'],
    [0.5, '1/2'],
    [0.625, '5/8'],
    [0.666, '2/3'],
    [0.75, '3/4'],
    [0.875, '7/8'],
  ];

  for (const [value, fraction] of fractions) {
    if (Math.abs(decimal - value) < 0.05) {
      return fraction;
    }
  }

  return decimal.toFixed(2);
}

// Common ingredient substitutions
export const INGREDIENT_SUBSTITUTIONS: Record<string, { substitute: string; ratio: string; notes: string }[]> = {
  'butter': [
    { substitute: 'coconut oil', ratio: '1:1', notes: 'Use refined for neutral flavor' },
    { substitute: 'olive oil', ratio: '3/4 cup per 1 cup butter', notes: 'Best for savory dishes' },
    { substitute: 'applesauce', ratio: '1:1', notes: 'For baking, reduces fat' },
    { substitute: 'greek yogurt', ratio: '1/2 cup per 1 cup butter', notes: 'Adds moisture and protein' },
  ],
  'egg': [
    { substitute: 'flax egg (1 tbsp ground flax + 3 tbsp water)', ratio: '1:1', notes: 'Let sit 5 mins to gel' },
    { substitute: 'chia egg (1 tbsp chia + 3 tbsp water)', ratio: '1:1', notes: 'Works great in brownies' },
    { substitute: 'mashed banana', ratio: '1/4 cup per egg', notes: 'Adds sweetness' },
    { substitute: 'applesauce', ratio: '1/4 cup per egg', notes: 'Good for moist baked goods' },
  ],
  'milk': [
    { substitute: 'oat milk', ratio: '1:1', notes: 'Creamy, great for baking' },
    { substitute: 'almond milk', ratio: '1:1', notes: 'Light, slightly nutty' },
    { substitute: 'coconut milk', ratio: '1:1', notes: 'Rich, adds tropical flavor' },
    { substitute: 'soy milk', ratio: '1:1', notes: 'Most similar to dairy milk' },
  ],
  'heavy cream': [
    { substitute: 'coconut cream', ratio: '1:1', notes: 'Rich and creamy' },
    { substitute: 'cashew cream', ratio: '1:1', notes: 'Blend soaked cashews with water' },
    { substitute: 'evaporated milk', ratio: '1:1', notes: 'Less rich but works well' },
  ],
  'flour': [
    { substitute: 'almond flour', ratio: '1:1', notes: 'Gluten-free, adds nuttiness' },
    { substitute: 'oat flour', ratio: '1:1', notes: 'Make by blending oats' },
    { substitute: 'coconut flour', ratio: '1/4 cup per 1 cup flour', notes: 'Very absorbent, add more liquid' },
  ],
  'sugar': [
    { substitute: 'honey', ratio: '3/4 cup per 1 cup sugar', notes: 'Reduce liquid in recipe slightly' },
    { substitute: 'maple syrup', ratio: '3/4 cup per 1 cup sugar', notes: 'Reduce liquid, adds flavor' },
    { substitute: 'coconut sugar', ratio: '1:1', notes: 'Lower glycemic index' },
    { substitute: 'stevia', ratio: '1 tsp per 1 cup sugar', notes: 'Very concentrated' },
  ],
  'soy sauce': [
    { substitute: 'coconut aminos', ratio: '1:1', notes: 'Slightly sweeter, soy-free' },
    { substitute: 'tamari', ratio: '1:1', notes: 'Gluten-free soy sauce' },
    { substitute: 'worcestershire sauce', ratio: '1:1', notes: 'Different flavor profile' },
  ],
  'chicken broth': [
    { substitute: 'vegetable broth', ratio: '1:1', notes: 'Vegetarian option' },
    { substitute: 'mushroom broth', ratio: '1:1', notes: 'Rich umami flavor' },
    { substitute: 'water + bouillon', ratio: '1 cup water + 1 tsp bouillon', notes: 'Quick substitute' },
  ],
  'lemon juice': [
    { substitute: 'lime juice', ratio: '1:1', notes: 'Slightly different citrus flavor' },
    { substitute: 'white wine vinegar', ratio: '1/2 the amount', notes: 'More acidic' },
    { substitute: 'apple cider vinegar', ratio: '1/2 the amount', notes: 'Adds fruity acidity' },
  ],
  'garlic': [
    { substitute: 'garlic powder', ratio: '1/8 tsp per clove', notes: 'Less pungent' },
    { substitute: 'shallots', ratio: '1/2 shallot per clove', notes: 'Milder, sweeter' },
    { substitute: 'garlic-infused oil', ratio: '1/2 tsp per clove', notes: 'For garlic flavor without pieces' },
  ],
};

// Find substitutions for an ingredient
export function findSubstitutions(ingredient: string): { substitute: string; ratio: string; notes: string }[] {
  const lowerIngredient = ingredient.toLowerCase();

  for (const [key, subs] of Object.entries(INGREDIENT_SUBSTITUTIONS)) {
    if (lowerIngredient.includes(key)) {
      return subs;
    }
  }

  return [];
}

// Estimate nutritional info (rough estimates per serving)
export interface NutritionEstimate {
  calories: number;
  protein: number; // grams
  carbs: number; // grams
  fat: number; // grams
  fiber: number; // grams
}

// Very rough estimates based on common ingredients
const INGREDIENT_NUTRITION: Record<string, Partial<NutritionEstimate>> = {
  'chicken': { calories: 165, protein: 31, fat: 3.6 },
  'beef': { calories: 250, protein: 26, fat: 15 },
  'salmon': { calories: 208, protein: 20, fat: 13 },
  'rice': { calories: 130, carbs: 28, fiber: 0.4 },
  'pasta': { calories: 131, carbs: 25, protein: 5 },
  'potato': { calories: 77, carbs: 17, fiber: 2.2 },
  'egg': { calories: 78, protein: 6, fat: 5 },
  'butter': { calories: 102, fat: 12 },
  'oil': { calories: 120, fat: 14 },
  'cheese': { calories: 113, protein: 7, fat: 9 },
  'bread': { calories: 79, carbs: 15, protein: 3 },
  'vegetable': { calories: 25, carbs: 5, fiber: 2 },
  'fruit': { calories: 60, carbs: 15, fiber: 2 },
  'bean': { calories: 120, protein: 8, carbs: 21, fiber: 7 },
  'lentil': { calories: 115, protein: 9, carbs: 20, fiber: 8 },
  'tofu': { calories: 80, protein: 8, fat: 4 },
  'milk': { calories: 103, protein: 8, carbs: 12, fat: 2 },
  'cream': { calories: 340, fat: 36 },
  'sugar': { calories: 49, carbs: 13 },
  'flour': { calories: 110, carbs: 23, protein: 3 },
};

export function estimateNutrition(recipe: Recipe): NutritionEstimate {
  let total: NutritionEstimate = {
    calories: 0,
    protein: 0,
    carbs: 0,
    fat: 0,
    fiber: 0,
  };

  recipe.ingredients.forEach((ing) => {
    const ingLower = ing.name.toLowerCase();

    for (const [key, nutrition] of Object.entries(INGREDIENT_NUTRITION)) {
      if (ingLower.includes(key)) {
        const amount = parseFloat(ing.amount) || 1;
        const multiplier = amount * (ing.unit.includes('cup') ? 1 : 0.5);

        total.calories += (nutrition.calories || 0) * multiplier;
        total.protein += (nutrition.protein || 0) * multiplier;
        total.carbs += (nutrition.carbs || 0) * multiplier;
        total.fat += (nutrition.fat || 0) * multiplier;
        total.fiber += (nutrition.fiber || 0) * multiplier;
        break;
      }
    }
  });

  // Divide by servings
  const servings = recipe.servings || 1;
  return {
    calories: Math.round(total.calories / servings),
    protein: Math.round(total.protein / servings),
    carbs: Math.round(total.carbs / servings),
    fat: Math.round(total.fat / servings),
    fiber: Math.round(total.fiber / servings),
  };
}

// Cooking timer presets
export interface TimerPreset {
  name: string;
  duration: number; // seconds
  icon: string;
  category: string;
}

export const TIMER_PRESETS: TimerPreset[] = [
  // Eggs
  { name: 'Soft boiled egg', duration: 360, icon: '🥚', category: 'Eggs' },
  { name: 'Medium boiled egg', duration: 480, icon: '🥚', category: 'Eggs' },
  { name: 'Hard boiled egg', duration: 600, icon: '🥚', category: 'Eggs' },

  // Pasta
  { name: 'Al dente pasta', duration: 540, icon: '🍝', category: 'Pasta' },
  { name: 'Well-done pasta', duration: 660, icon: '🍝', category: 'Pasta' },

  // Rice
  { name: 'White rice', duration: 1080, icon: '🍚', category: 'Grains' },
  { name: 'Brown rice', duration: 2700, icon: '🍚', category: 'Grains' },

  // Proteins
  { name: 'Chicken breast (each side)', duration: 420, icon: '🍗', category: 'Protein' },
  { name: 'Steak medium-rare (each side)', duration: 240, icon: '🥩', category: 'Protein' },
  { name: 'Salmon fillet', duration: 480, icon: '🐟', category: 'Protein' },

  // Baking
  { name: 'Cookies', duration: 720, icon: '🍪', category: 'Baking' },
  { name: 'Brownies', duration: 1500, icon: '🍫', category: 'Baking' },
  { name: 'Cake', duration: 1800, icon: '🎂', category: 'Baking' },
  { name: 'Bread loaf', duration: 2400, icon: '🍞', category: 'Baking' },

  // Vegetables
  { name: 'Roasted vegetables', duration: 1800, icon: '🥕', category: 'Vegetables' },
  { name: 'Steamed broccoli', duration: 300, icon: '🥦', category: 'Vegetables' },
  { name: 'Baked potato', duration: 3600, icon: '🥔', category: 'Vegetables' },
];

// Convert unit helpers
export const UNIT_CONVERSIONS: Record<string, { to: string; factor: number }[]> = {
  'cup': [
    { to: 'ml', factor: 236.588 },
    { to: 'tbsp', factor: 16 },
    { to: 'tsp', factor: 48 },
  ],
  'tbsp': [
    { to: 'tsp', factor: 3 },
    { to: 'ml', factor: 14.787 },
    { to: 'cup', factor: 0.0625 },
  ],
  'tsp': [
    { to: 'ml', factor: 4.929 },
    { to: 'tbsp', factor: 0.333 },
  ],
  'oz': [
    { to: 'g', factor: 28.35 },
    { to: 'lb', factor: 0.0625 },
  ],
  'lb': [
    { to: 'g', factor: 453.592 },
    { to: 'kg', factor: 0.453592 },
    { to: 'oz', factor: 16 },
  ],
  'g': [
    { to: 'oz', factor: 0.035274 },
    { to: 'kg', factor: 0.001 },
  ],
};

export function convertUnit(amount: number, fromUnit: string, toUnit: string): number | null {
  const conversions = UNIT_CONVERSIONS[fromUnit.toLowerCase()];
  if (!conversions) return null;

  const conversion = conversions.find((c) => c.to === toUnit.toLowerCase());
  if (!conversion) return null;

  return amount * conversion.factor;
}

// Meal planning helper
export interface MealPlan {
  id: string;
  name: string;
  days: {
    day: string;
    meals: {
      type: 'breakfast' | 'lunch' | 'dinner' | 'snack';
      recipeId?: string;
      recipeName?: string;
      notes?: string;
    }[];
  }[];
  createdAt: number;
}

export function createMealPlan(name: string): MealPlan {
  const days = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'];

  return {
    id: `plan-${Date.now()}`,
    name,
    days: days.map((day) => ({
      day,
      meals: [
        { type: 'breakfast' },
        { type: 'lunch' },
        { type: 'dinner' },
      ],
    })),
    createdAt: Date.now(),
  };
}

// Get shopping list from meal plan
export function getShoppingListFromMealPlan(mealPlan: MealPlan, recipes: Recipe[]): string[] {
  const ingredients = new Map<string, string>();

  mealPlan.days.forEach((day) => {
    day.meals.forEach((meal) => {
      if (meal.recipeId) {
        const recipe = recipes.find((r) => r.id === meal.recipeId);
        if (recipe) {
          recipe.ingredients.forEach((ing) => {
            const key = ing.name.toLowerCase();
            if (ingredients.has(key)) {
              // Try to combine amounts (simplified)
              const existing = ingredients.get(key)!;
              ingredients.set(key, `${existing} + ${ing.amount} ${ing.unit}`);
            } else {
              ingredients.set(key, `${ing.amount} ${ing.unit} ${ing.name}`);
            }
          });
        }
      }
    });
  });

  return Array.from(ingredients.values());
}

File: lib/core/services/cooking_helpers.dart
import '../../shared/models/models.dart';

// --- Recipe Scaling ---

Recipe scaleRecipe(Recipe recipe, int newServings) {
  if (recipe.servings == newServings) return recipe;

  final factor = newServings / recipe.servings;

  final scaledIngredients = recipe.ingredients.map((ing) {
    final parsed = double.tryParse(ing.amount);
    if (parsed == null) return ing;

    final scaled = parsed * factor;
    final formatted = scaled < 1
        ? _toFraction(scaled)
        : scaled == scaled.roundToDouble()
            ? scaled.toInt().toString()
            : scaled.toStringAsFixed(1);

    return ing.copyWith(amount: formatted);
  }).toList();

  return recipe.copyWith(
    servings: newServings,
    ingredients: scaledIngredients,
  );
}

String _toFraction(double val) {
  if ((val - 0.25).abs() < 0.05) return '1/4';
  if ((val - 0.33).abs() < 0.05) return '1/3';
  if ((val - 0.5).abs() < 0.05) return '1/2';
  if ((val - 0.67).abs() < 0.05) return '2/3';
  if ((val - 0.75).abs() < 0.05) return '3/4';
  return val.toStringAsFixed(2);
}

// --- Ingredient Substitutions ---

class Substitution {
  final String name;
  final String ratio;
  final String notes;

  const Substitution({
    required this.name,
    required this.ratio,
    required this.notes,
  });
}

const ingredientSubstitutions = <String, List<Substitution>>{
  'butter': [
    Substitution(name: 'Coconut Oil', ratio: '1:1', notes: 'Same amount, solid form'),
    Substitution(name: 'Olive Oil', ratio: '3/4 cup per 1 cup', notes: 'Best for savory'),
    Substitution(name: 'Applesauce', ratio: '1:1', notes: 'For baking, reduces fat'),
    Substitution(name: 'Greek Yogurt', ratio: '1:1', notes: 'Adds moisture and tang'),
  ],
  'egg': [
    Substitution(name: 'Flax Egg', ratio: '1 tbsp ground flax + 3 tbsp water', notes: 'Let sit 5 min'),
    Substitution(name: 'Chia Egg', ratio: '1 tbsp chia + 3 tbsp water', notes: 'Let sit 5 min'),
    Substitution(name: 'Mashed Banana', ratio: '1/4 cup per egg', notes: 'Adds sweetness'),
    Substitution(name: 'Applesauce', ratio: '1/4 cup per egg', notes: 'Neutral flavor'),
  ],
  'milk': [
    Substitution(name: 'Oat Milk', ratio: '1:1', notes: 'Creamy, neutral'),
    Substitution(name: 'Almond Milk', ratio: '1:1', notes: 'Lighter, nutty'),
    Substitution(name: 'Coconut Milk', ratio: '1:1', notes: 'Rich, sweet'),
    Substitution(name: 'Soy Milk', ratio: '1:1', notes: 'High protein'),
  ],
  'heavy cream': [
    Substitution(name: 'Coconut Cream', ratio: '1:1', notes: 'Thick and rich'),
    Substitution(name: 'Cashew Cream', ratio: '1:1', notes: 'Blend soaked cashews'),
    Substitution(name: 'Evaporated Milk', ratio: '1:1', notes: 'Lower fat'),
  ],
  'flour': [
    Substitution(name: 'Almond Flour', ratio: '1:1', notes: 'Gluten-free, nutty'),
    Substitution(name: 'Oat Flour', ratio: '1:1', notes: 'Blend rolled oats'),
    Substitution(name: 'Coconut Flour', ratio: '1/4 cup per 1 cup', notes: 'Very absorbent'),
  ],
  'sugar': [
    Substitution(name: 'Honey', ratio: '3/4 cup per 1 cup', notes: 'Reduce other liquids'),
    Substitution(name: 'Maple Syrup', ratio: '3/4 cup per 1 cup', notes: 'Distinct flavor'),
    Substitution(name: 'Coconut Sugar', ratio: '1:1', notes: 'Lower glycemic'),
    Substitution(name: 'Stevia', ratio: '1 tsp per 1 cup', notes: 'Very concentrated'),
  ],
  'soy sauce': [
    Substitution(name: 'Coconut Aminos', ratio: '1:1', notes: 'Soy-free, sweeter'),
    Substitution(name: 'Tamari', ratio: '1:1', notes: 'Gluten-free soy sauce'),
    Substitution(name: 'Worcestershire', ratio: '1:1', notes: 'Different flavor profile'),
  ],
  'chicken broth': [
    Substitution(name: 'Vegetable Broth', ratio: '1:1', notes: 'Vegetarian option'),
    Substitution(name: 'Mushroom Broth', ratio: '1:1', notes: 'Rich umami'),
    Substitution(name: 'Bouillon + Water', ratio: '1 cube per cup', notes: 'Quick option'),
  ],
  'lemon juice': [
    Substitution(name: 'Lime Juice', ratio: '1:1', notes: 'Similar acidity'),
    Substitution(name: 'White Wine Vinegar', ratio: '1/2 amount', notes: 'Stronger acid'),
    Substitution(name: 'Apple Cider Vinegar', ratio: '1/2 amount', notes: 'Fruity tang'),
  ],
  'garlic': [
    Substitution(name: 'Garlic Powder', ratio: '1/4 tsp per clove', notes: 'Less pungent'),
    Substitution(name: 'Shallots', ratio: '1 shallot per 3 cloves', notes: 'Milder'),
    Substitution(name: 'Garlic-infused Oil', ratio: '1/2 tsp per clove', notes: 'Low-FODMAP'),
  ],
};

List<Substitution> findSubstitutions(String ingredient) {
  final lower = ingredient.toLowerCase();
  for (final entry in ingredientSubstitutions.entries) {
    if (lower.contains(entry.key)) return entry.value;
  }
  return [];
}

// --- Nutrition Estimation ---

const _nutritionPerUnit = <String, NutritionEstimate>{
  'chicken': NutritionEstimate(calories: 165, protein: 31, carbs: 0, fat: 4, fiber: 0),
  'beef': NutritionEstimate(calories: 250, protein: 26, carbs: 0, fat: 17, fiber: 0),
  'rice': NutritionEstimate(calories: 206, protein: 4, carbs: 45, fat: 0, fiber: 1),
  'pasta': NutritionEstimate(calories: 220, protein: 8, carbs: 43, fat: 1, fiber: 3),
  'egg': NutritionEstimate(calories: 72, protein: 6, carbs: 0, fat: 5, fiber: 0),
  'butter': NutritionEstimate(calories: 102, protein: 0, carbs: 0, fat: 12, fiber: 0),
  'oil': NutritionEstimate(calories: 120, protein: 0, carbs: 0, fat: 14, fiber: 0),
  'cheese': NutritionEstimate(calories: 113, protein: 7, carbs: 0, fat: 9, fiber: 0),
  'milk': NutritionEstimate(calories: 149, protein: 8, carbs: 12, fat: 8, fiber: 0),
  'cream': NutritionEstimate(calories: 340, protein: 3, carbs: 3, fat: 36, fiber: 0),
  'vegetable': NutritionEstimate(calories: 50, protein: 2, carbs: 10, fat: 0, fiber: 3),
  'fruit': NutritionEstimate(calories: 60, protein: 1, carbs: 15, fat: 0, fiber: 2),
  'bean': NutritionEstimate(calories: 120, protein: 8, carbs: 22, fat: 0, fiber: 6),
  'bread': NutritionEstimate(calories: 79, protein: 3, carbs: 15, fat: 1, fiber: 1),
};

NutritionEstimate estimateNutrition(Recipe recipe) {
  int totalCal = 0, totalPro = 0, totalCarb = 0, totalFat = 0, totalFib = 0;

  for (final ing in recipe.ingredients) {
    final lower = ing.name.toLowerCase();
    for (final entry in _nutritionPerUnit.entries) {
      if (lower.contains(entry.key)) {
        final multiplier = _unitMultiplier(ing.unit);
        totalCal += (entry.value.calories * multiplier).round();
        totalPro += (entry.value.protein * multiplier).round();
        totalCarb += (entry.value.carbs * multiplier).round();
        totalFat += (entry.value.fat * multiplier).round();
        totalFib += (entry.value.fiber * multiplier).round();
        break;
      }
    }
  }

  final servings = recipe.servings > 0 ? recipe.servings : 1;
  return NutritionEstimate(
    calories: totalCal ~/ servings,
    protein: totalPro ~/ servings,
    carbs: totalCarb ~/ servings,
    fat: totalFat ~/ servings,
    fiber: totalFib ~/ servings,
  );
}

double _unitMultiplier(String unit) {
  final lower = unit.toLowerCase();
  if (lower.contains('cup')) return 1.0;
  if (lower.contains('tbsp')) return 0.25;
  if (lower.contains('tsp')) return 0.1;
  return 0.5;
}

// --- Timer Presets ---

class TimerPreset {
  final String name;
  final String category;
  final int minutes;

  const TimerPreset({
    required this.name,
    required this.category,
    required this.minutes,
  });
}

const timerPresets = <TimerPreset>[
  // Eggs
  TimerPreset(name: 'Soft Boiled Egg', category: 'Eggs', minutes: 6),
  TimerPreset(name: 'Medium Boiled Egg', category: 'Eggs', minutes: 8),
  TimerPreset(name: 'Hard Boiled Egg', category: 'Eggs', minutes: 10),
  // Pasta
  TimerPreset(name: 'Pasta Al Dente', category: 'Pasta', minutes: 9),
  TimerPreset(name: 'Pasta Well Done', category: 'Pasta', minutes: 11),
  // Grains
  TimerPreset(name: 'White Rice', category: 'Grains', minutes: 18),
  TimerPreset(name: 'Brown Rice', category: 'Grains', minutes: 45),
  // Proteins
  TimerPreset(name: 'Chicken Breast', category: 'Proteins', minutes: 8),
  TimerPreset(name: 'Steak (Medium)', category: 'Proteins', minutes: 4),
  TimerPreset(name: 'Salmon Fillet', category: 'Proteins', minutes: 5),
  // Baking
  TimerPreset(name: 'Cookies', category: 'Baking', minutes: 12),
  TimerPreset(name: 'Brownies', category: 'Baking', minutes: 25),
  TimerPreset(name: 'Cake', category: 'Baking', minutes: 30),
  TimerPreset(name: 'Bread', category: 'Baking', minutes: 40),
  // Vegetables
  TimerPreset(name: 'Roasted Vegetables', category: 'Vegetables', minutes: 25),
  TimerPreset(name: 'Steamed Vegetables', category: 'Vegetables', minutes: 5),
];

// --- Unit Conversions ---

class UnitConversion {
  final String from;
  final String to;
  final double factor;

  const UnitConversion({
    required this.from,
    required this.to,
    required this.factor,
  });
}

const unitConversions = <UnitConversion>[
  UnitConversion(from: 'cup', to: 'ml', factor: 236.588),
  UnitConversion(from: 'cup', to: 'tbsp', factor: 16),
  UnitConversion(from: 'cup', to: 'tsp', factor: 48),
  UnitConversion(from: 'tbsp', to: 'tsp', factor: 3),
  UnitConversion(from: 'tbsp', to: 'ml', factor: 14.787),
  UnitConversion(from: 'tsp', to: 'ml', factor: 4.929),
  UnitConversion(from: 'oz', to: 'g', factor: 28.3495),
  UnitConversion(from: 'lb', to: 'g', factor: 453.592),
  UnitConversion(from: 'lb', to: 'kg', factor: 0.4536),
  UnitConversion(from: 'oz', to: 'ml', factor: 29.5735),
];

double? convertUnit(double value, String from, String to) {
  if (from == to) return value;

  for (final c in unitConversions) {
    if (c.from == from && c.to == to) return value * c.factor;
    if (c.from == to && c.to == from) return value / c.factor;
  }
  return null;
}

// --- Meal Plan Helpers ---

MealPlan createEmptyMealPlan() {
  final now = DateTime.now();
  final monday = now.subtract(Duration(days: now.weekday - 1));

  final days = List.generate(7, (i) {
    final date = monday.add(Duration(days: i));
    const dayNames = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'];
    return MealDay(
      dayName: dayNames[i],
      date: date,
      meals: [
        const MealSlot(type: MealType.breakfast),
        const MealSlot(type: MealType.lunch),
        const MealSlot(type: MealType.dinner),
        const MealSlot(type: MealType.snack),
      ],
    );
  });

  return MealPlan(days: days);
}

List<Ingredient> getShoppingListFromMealPlan(MealPlan plan) {
  final ingredients = <Ingredient>[];
  for (final day in plan.days) {
    for (final slot in day.meals) {
      if (slot.recipe != null) {
        ingredients.addAll(slot.recipe!.ingredients);
      }
    }
  }
  return ingredients;
}

// Health goal daily calorie targets
const healthGoalCalories = <HealthGoal, int>{
  HealthGoal.weightLoss: 1500,
  HealthGoal.muscleBuilding: 2500,
  HealthGoal.balanced: 2000,
  HealthGoal.lowCarb: 1800,
  HealthGoal.heartHealthy: 1800,
};
